### PR TITLE
Always always always throw/reject with an Error

### PR DESCRIPTION
**Pull request by jridgewell**
Sorry, this one changes a lot of areas. Please review that my assumptions are correct (that we are indeed throwing an Error instance), trace it if you have to.

When reviewing, please review 9129ef8. Except @erwinmombay, I need you to review 74435e5, too.

/to @tdrl: Please review A4A and `bytes.js` changes
/to @lannka: Please review A4A changes
/to @dvoytenko: Please review `<amp-access>`, `animation.js`, `motion.js`, `extensions-impl.js`, all the resources, and viewer changes. 😬
/to @aghassemi: Please review `<amp-viz-vega>` changes
/to @cramforce: Please review `chunk.js` changes

### CODE REVIEW

File: src/chunk.js [hunk 1]
```diff
@@ -71,18 +71,18 @@ export function activateChunkingForTesting() {
...
+function runChunks(service) {
+  try {
+    while (service.execute_()) {
+      // noop.
     }
-  }
-  if (errors.length) {
-    throw errors[0];
+  } catch (e) {
+    // Can you feel the recursion!
```
Comment (cramforce) [inline src/chunk.js]: Oh, wow.

File: src/utils/bytes.js [hunk 2]
```diff
@@ -24,12 +24,14 @@ import {dev} from '../log';
  */
 export function utf8Decode(bytes) {
   if (TextDecoder) {
-    return Promise.resolve(new TextDecoder('utf-8').decode(bytes));
+    return new Promise(resolve => {
```
Comment (tdrl) [inline src/utils/bytes.js]: Just curious: Why is it useful to inject a layer of deferred execution here?

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -501,9 +500,9 @@ export class AmpA4A extends AMP.BaseElement {
         }
       }
       if (rendered instanceof Error) {
-        throw rendered;
+        throw dev().assertError(rendered);
```
Comment (tdrl) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Is the `assertError` useful here?  We've already verified that it's an `Error`.

File: extensions/amp-access/0.1/login-dialog.js [hunk 4]
```diff
@@ -267,7 +267,7 @@ export class WebLoginDialog {
     }
     dev().fine(TAG, 'Login done: ', result, opt_error);
     if (opt_error) {
-      this.reject_(opt_error);
+      this.reject_(user().createError('login failed', opt_error));
```
Comment (dvoytenko) [inline extensions/amp-access/0.1/login-dialog.js]: Why? Isn't error already marked correctly as a user error or dev error? Why are we changing the meaning of the error?

File: extensions/amp-form/0.1/amp-form.js [hunk 5]
```diff
@@ -209,7 +209,7 @@ export class AmpForm {
         this.actions_.trigger(this.form_, 'submit-error', null);
         this.setState_(FormState_.SUBMIT_ERROR);
         this.renderTemplate_(error.responseJson || {});
-        rethrowAsync('Form submission failed:', error);
+        rethrowAsync(user().createError('Form submission failed:', error));
```
Comment (dvoytenko) [inline extensions/amp-form/0.1/amp-form.js]: Hmm. Normally a point of `rethrowAsync` is to preserve the user/dev flag. This might be correct, but to confirm: why are we converting this error to a user error?

File: extensions/amp-app-banner/0.1/amp-app-banner.js [hunk 6]
```diff
@@ -396,7 +396,7 @@ export class AmpAndroidAppBanner extends AbstractAppBanner {
         .then(response => this.parseManifest_(response))
         .catch(error => {
           this.hide_();
-          rethrowAsync(error);
+          rethrowAsync(dev().assertError(error));
```
Comment (dvoytenko) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Would it be simpler if we just "asserted the error" inside `rethrowAsync` call?

File: src/animation.js [hunk 7]
```diff
@@ -281,17 +278,19 @@ class AnimationPlayer {
           }
         }
       } catch (e) {
-        dev().error(TAG_, 'completion failed: ' + e, e);
-        success = false;
+        error = dev().createError('completion failed', e);
```
Comment (dvoytenko) [inline src/animation.js]: Any error is by default a dev error. Why clarify it multiple times?

File: src/animation.js [hunk 8]
```diff
@@ -281,17 +278,19 @@ class AnimationPlayer {
...
+        dev().error(TAG_, error);
       }
     }
-    if (success) {
-      this.resolve_();
+    if (error) {
+      this.reject_(dev().assertError(error));
```
Comment (dvoytenko) [inline src/animation.js]: Do we still have problem here with `if (x) {assert(x)}`?

File: src/animation.js [hunk 9]
```diff
@@ -323,13 +322,13 @@ class AnimationPlayer {
...
     if (normLinearTime == 1) {
-      this.complete_(/* success */ true, /* dir */ 0);
+      this.complete_(null, /* dir */ 0);
     } else {
       if (this.vsync_.canAnimate(this.contextNode_)) {
         this.task_(this.state_);
       } else {
         dev().warn(TAG_, 'cancel animation');
-        this.complete_(/* success */ false, /* dir */ 0);
+        this.complete_(new Error('cancel animation'), /* dir */ 0);
```
Comment (dvoytenko) [inline src/animation.js]: Aren't we supposed to use `cancellation()` from `error.js` in these cases?

File: src/motion.js [hunk 10]
```diff
@@ -174,19 +175,17 @@ export class Motion {
    */
   halt() {
     if (this.continuing_) {
-      this.completeContinue_(false);
+      this.completeContinue_(new Error('halting motion'));
```
Comment (dvoytenko) [inline src/motion.js]: Ditto: `cancelation()`

File: src/service/extensions-impl.js [hunk 11]
```diff
@@ -314,14 +314,16 @@ export class Extensions {
           try {
             factory(/** @type {!ShadowRoot} */ (ampdoc.getRootNode()));
           } catch (e) {
-            rethrowAsync('ShadowRoot factory failed: ', e, extensionId);
+            rethrowAsync(dev().createError('ShadowRoot factory failed: ',
```
Comment (dvoytenko) [inline src/service/extensions-impl.js]: Ditto: we are re-marking a potential "user" error as "dev" error here. This is obviously not very likely during factory installation, but then why risk it?

File: src/service/resource.js [hunk 12]
```diff
@@ -614,7 +614,7 @@ export class Resource {
       dev().fine(TAG, 'layout complete:', this.debugid);
     } else {
       dev().fine(TAG, 'loading failed:', this.debugid, opt_reason);
-      return Promise.reject(opt_reason);
+      return Promise.reject(dev().createError('loading failed', opt_reason));
```
Comment (dvoytenko) [inline src/service/resource.js]: Ditto: we just turned a user error into a dev error.

File: src/utils/bytes.js (hunk 2 above)
Comment (jridgewell) [inline src/utils/bytes.js]: `TextDecoder#decode` doesn't return a promise, and it throws synchronously on errors, which is not the intended behavior for `utf8Decode`. Also, promise resolvers (the function passed to the constructor) as executed sync, so it's not deferred. But, the resolver is wrapped in a try-catch in the internal Promise code, so if it throws sync it will result in the new promise being rejected.

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Ideally we wouldn't need it, but the linting system can only account for so many circumstances. The `dev().assertError()` is an easy way for the linter to ensure this is really an error instance. Plus, `dev().assertError(rendered)` turns into `rendered` when we compile the production code.

File: extensions/amp-access/0.1/login-dialog.js (hunk 4 above)
Comment (jridgewell) [inline extensions/amp-access/0.1/login-dialog.js]: From what I can find, the only places this error can be are:

- https://github.com/jridgewell/amphtml/blob/ccbe994f5372b65173e1d2d224c74ef3cc6a07a6/extensions/amp-access/0.1/login-dialog.js#L203
- https://github.com/jridgewell/amphtml/blob/ccbe994f5372b65173e1d2d224c74ef3cc6a07a6/extensions/amp-access/0.1/login-dialog.js#L215

It could also be from a [rejected `#urlOrPromise`](https://github.com/jridgewell/amphtml/blob/ccbe994f5372b65173e1d2d224c74ef3cc6a07a6/extensions/amp-access/0.1/login-dialog.js#L198), which can never happen. So it's always an unmarked Error instance. I could change this to `dev().assertError()`, but then we'd lose the `"login failed"` context in the error message.

File: extensions/amp-app-banner/0.1/amp-app-banner.js (hunk 6 above)
Comment (jridgewell) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: We'd lose the typing from closure then?

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: @mkhatib could answer this better, but I think any error here is user related, as the errors come from `#fetchJSON`. It's possible the error could be from a sync exception in the `#then` block directly before this `#catch` (which is definitely a dev error), but we could avoid that by moving this block to be the second parameter to the `#then` above.

File: extensions/amp-viz-vega/0.1/amp-viz-vega.js [hunk 13]
```diff
@@ -193,7 +193,7 @@ export class AmpVizVega extends AMP.BaseElement {
     const parsePromise = new Promise((resolve, reject) => {
       this.vega_.parse.spec(this.data_, (error, chartFactory) => {
         if (error) {
-          reject(error);
+          reject(dev().createError('parsing failed', error));
```
Comment (jridgewell) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: @aghassemi: should this be a user or dev error?

File: src/animation.js (hunk 7 above)
Comment (jridgewell) [inline src/animation.js]: I'm adding the additional `completion failed` context to the error message.

File: src/animation.js (hunk 8 above)
Comment (jridgewell) [inline src/animation.js]: Hm?

File: src/animation.js (hunk 9 above)
Comment (jridgewell) [inline src/animation.js]: I thought those were for XHR requests? I'd be open to using it here.

File: src/service/resource.js (hunk 12 above)
Comment (jridgewell) [inline src/service/resource.js]: Same reasoning, the layout of individual elements seems like a dev concern.

File: extensions/amp-viz-vega/0.1/amp-viz-vega.js (hunk 13 above)
Comment (aghassemi) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: user (ie. their spec JSON is invalid)

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 14]
```diff
@@ -468,7 +467,7 @@ export class AmpA4A extends AMP.BaseElement {
       'au': adQueryIdx < 0 ? '' :
           this.adUrl_.substring(adQueryIdx + 1, adQueryIdx + 251),
     };
-    return new Error('amp-a4a: ' + JSON.stringify(state));
+    throw new Error('amp-a4a: ' + JSON.stringify(state));
```
Comment (lannka) [inline extensions/amp-a4a/0.1/amp-a4a.js]: is this change intentional?

File: extensions/amp-access/0.1/login-dialog.js (hunk 4 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/login-dialog.js]: Seems a bit brittle w.r.t. future changes. I'm not sure how valuable "login failed" since we do have error stack. If it is, maybe something like the current `createErrorVargs` in `log.js` better? It extends the error message but preserves user/dev flag.

File: src/animation.js (hunk 7 above)
Comment (dvoytenko) [inline src/animation.js]: Ok. Cool.

File: src/animation.js [hunk 15]
```diff
@@ -281,17 +278,19 @@ class AnimationPlayer {
           }
         }
       } catch (e) {
-        dev().error(TAG_, 'completion failed: ' + e, e);
-        success = false;
+        error = dev().createError('completion failed', e);
+        dev().error(TAG_, error);
```
Comment (dvoytenko) [inline src/animation.js]: At this point it's probably just better to call rethrowAsync(error)

File: src/animation.js (hunk 9 above)
Comment (dvoytenko) [inline src/animation.js]: I believe the intent is to simply give more info to our loggers and error reporting that this is not a "real" error, but rather an "abort" condition.

File: src/service/resource.js (hunk 12 above)
Comment (dvoytenko) [inline src/service/resource.js]: Many layout problems could be a result of "user" error, such as misconfiguring an AMP element or problems executing XHRs (404, etc). I think this is a prevalent case actually.

File: extensions/amp-access/0.1/login-dialog.js (hunk 4 above)
Comment (jridgewell) [inline extensions/amp-access/0.1/login-dialog.js]: Chances are that these are unmarked errors (the two above), and they are definitely user errors and not dev errors. How about making it so `Log#createError` doesn't change the user/dev mark if an error passed in already has one?

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (jridgewell) [inline extensions/amp-form/0.1/amp-form.js]: Let's have this conversation here: https://github.com/ampproject/amphtml/pull/5841#discussion_r85375956

File: extensions/amp-app-banner/0.1/amp-app-banner.js (hunk 6 above)
Comment (jridgewell) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Can you explain?

File: src/animation.js (hunk 8 above)
Comment (jridgewell) [inline src/animation.js]: Yes, but it's really hard to get that in the linting system. Just using `#assertError` takes care of like 10 different `if (error)`, `if (error instanceof Error)`, etc, cases.

File: src/animation.js (hunk 15 above)
Comment (jridgewell) [inline src/animation.js]: Done.

File: src/animation.js (hunk 9 above)
Comment (jridgewell) [inline src/animation.js]: Done.

File: src/motion.js (hunk 10 above)
Comment (jridgewell) [inline src/motion.js]: Done.

File: src/service/resource.js (hunk 12 above)
Comment (jridgewell) [inline src/service/resource.js]: Let's continue here: https://github.com/ampproject/amphtml/pull/5841#discussion_r85375956

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 16]
```diff
@@ -409,11 +409,11 @@ export class AmpA4A extends AMP.BaseElement {
                         if (isValid) {
                           return creativeParts.creative;
                         }
-                        return Promise.reject(
-                            'Key failed to validate creative\'s signature.');
+                        throw new Error("Key failed to validate creative's " +
```
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Let's discuss that in [another issue](https://github.com/ampproject/amphtml/issues/5889). For now, this keeps the same behavior.

Comment (tdrl) [inline extensions/amp-a4a/0.1/amp-a4a.js]: +1  We know that we need to clean up a lot of the logic in here and clarify what's a real error, what just needs a console message, etc.  But that's a bigger and more unified effort than this change.

File: extensions/amp-access/0.1/login-dialog.js (hunk 4 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/login-dialog.js]: Those both are `dev` errors. The point of `createError` at this time is to override the dev/user flag, though we could certainly have one (different one, not part of `dev` or `user`) that preserves it.

File: extensions/amp-app-banner/0.1/amp-app-banner.js (hunk 6 above)
Comment (dvoytenko) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: I guess I still don't understand why the new code is an improvement over the old one: it's certainly not easier to read and doesn't really give us more information.

**Comment by dvoytenko** [comment]
@jridgewell I still have concerns over this PR. It's a big change. I understand that there are (looks like very few) older and maybe newer cases where errors came without stack and these cases definitely need to be fixed. However, overall, I don't see an improvement over the previous state. The number of issues we have right now show that the "forgotten error stack" is definitely not a major issue in our code.

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 14 above)
Comment (lannka) [inline extensions/amp-a4a/0.1/amp-a4a.js]: agree that the current code needs to be improved. 

but line 453 still returns an error object though. do you want to fix that as well?

File: build-system/eslint-rules/reject-with-error.js [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+  });
+}
+
+function isLogCreateErrorCall(node) {
+  if (node.type != 'CallExpression') {
+    return false;
+  }
+
+  var call = node.callee;
+  return call.type == 'MemberExpression' &&
```
Comment (erwinmombay) [inline build-system/eslint-rules/reject-with-error.js]: not sure where my comments went. (think i commented last week)

can you update to either double or triple for consistency.

File: build-system/eslint-rules/reject-with-error.js [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+  if (call.property.name == 'catch') {
+    return true;
+  }
+
+  return false;
+}
+
+function parameterIsExplicitlyError(param) {
+  var func = param.parent;
+  var regex = new RegExp('@param {!(Type)?Error} ' + param.name);
```
Comment (erwinmombay) [inline build-system/eslint-rules/reject-with-error.js]: cool. so this looks at declarations of inline declarations right?

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 14 above)
Comment (jridgewell) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Good catch.

File: extensions/amp-app-banner/0.1/amp-app-banner.js (hunk 6 above)
Comment (jridgewell) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: It's necessary for the linter. It's easy to see a call to `#assertError` and know that the return value is definitely an Error instance.

File: extensions/amp-access/0.1/login-dialog.js (hunk 4 above)
Comment (jridgewell) [inline extensions/amp-access/0.1/login-dialog.js]: Ah, ok, I can change them to be dev errors then. But I'd like to see is something like a first mark survives. So, if something has been marked as a user error in some inner code, trying `dev().mabyeMark()` can still provide context but doesn't change the user mark.

**Comment by lannka** [comment]
@jridgewell is this still going to happen?

**Comment by jridgewell** [comment]
I would love for this to happen, but it'll take a lot of work.

**Comment by dvoytenko** [comment]
@erwinmombay Started to look how much of this can be addressed on the compiler level. E.g. prohibit to `throw` with non-error type and the same for `Promise.reject`.

**Comment by lannka** [comment]
@jridgewell  this PR is too old. there might be large merge conflicts. I'll close it for now, and you can reopen when you're back on it.
